### PR TITLE
Initial Specification, Now With Forward Secrecy:tm:

### PR DESCRIPTION
**Pull request by expede**
[👀📜 Preview](https://github.com/ucan-wg/awake/blob/port/README.md)

# TL;DR

An update to the AWAKE protocol — standardization between Fission and Bluesky 🤝 

# Changelog

* Generalized from secure login to any secure session
* WebCrypto-friendly Double Ratchet KDF
* Consistent message format
* More of the handshake is in the ciphertext payload
* Add spell checker
* Add markdown link checker

# Related GH Issues
* https://github.com/fission-codes/webnative/issues/337

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,316 @@
...
+* [Daniel Holmgren](https://github.com/dholms), [Bluesky](https://blueskyweb.xyz/)
+
+# 0. Abstract
+
+Authorized Wire Authenticated Key Exchange (AWAKE) is an [AKE](https://en.wikipedia.org/wiki/Authenticated_Key_Exchange) built on top of the [UCAN auth token](https://github.com/ucan-wg/spec). AWAKE is similar to other mutual authentication schemes, such as self-signed [mTLS](https://datatracker.ietf.org/doc/html/rfc8705), but with a focus on authorization proofs. AWAKE leverages the capability chain to prove access to some resource, validating that the requestor is communicating with a party capable of performing certain actions. This is a helpful root of trust with a well defined context when establishing a secure communications channel.
+
+The core problem that AWAKE solves is bootstrapping a secure session on top of a public channel. Key exchanges for point-to-point communication are plentiful, but in open, trusteless protocols, rooting trust can be a barrier for ad hoc communications channels. Two common approaches are to use a trusted certificate authority, or ignore the principal and "merely" establish a point-to-point channel.
```
Comment (bgins) [inline README.md:16]: Who is "the principal" here? Is it a certificate authority or some other root of trust?

File: README.md [hunk 2]
```diff
@@ -1 +1,316 @@
...
+
+Capability-based systems have a helpful philosophy towards a third path. By emphasizing authorization over authentication, they provide a way to know something provable about what the other party "can do", even if they have no sure way of knowing "who they are". One way of phrasing this is that such an agent is "functionally equivalent to the principal in this context".
```
Comment (bgins) [inline README.md:18]: Yes! The principal makes sense reading this section. It might be good to contextualize the term earlier in the document.

File: README.md [hunk 3]
```diff
@@ -1 +1,316 @@
...
+| Attacker  | An attacker attempting to gain access to the channel |
+
+## 2 Sequence
+
+AWAKE proceeds in one connecion step and four communication rounds:
+
+1. Both parties subscribe to a well-known channel
+2. Requester broadcasts intent
+    * a. Temporary DID
+    * b. Responder authorization criterea
```
Comment (bgins) [inline README.md:39]: >    * b. Responder authorization ~criterea~ criteria

Criteria, not criterea.

File: README.md [hunk 4]
```diff
@@ -1 +1,457 @@
...
+
+## Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+# 1 Introduction
+
+AWAKE bootstraps a secure session on top of a public channel. Key exchanges for point-to-point communication are plentiful, but in open, trusteless protocols, rooting trust can be a barrier for ad hoc communications channels. Two common approaches are to use a trusted certificate authority, or ignore the principal and "merely" establish a point-to-point channel.
```
Comment (dholms) [inline README.md:22]: sp: "trustless"

File: README.md [hunk 5]
```diff
@@ -1 +1,457 @@
...
+| --------- | ---------------------------------------------------- |
+| Requestor | The agent opening the session                        |
+| Responder | The agent being contacted by the Requestor           |
+| Attacker  | An attacker attempting to gain access to the channel |
+
+## 2 Sequence
+
+AWAKE proceeds in one connection step, four communication rounds, and an OPTIONAL disconnection:
+
+1. Both parties subscribe to a well-known channel
```
Comment (dholms) [inline README.md:49]: Is the channel assumed to be long standing?

If not, how often does the requestor broadcast intent?

If the provider joins _after_ the requestor broadcasts intent, when does it get the update that it needs to re-broadcast?

File: README.md [hunk 6]
```diff
@@ -1 +1,457 @@
...
+│  │  aud: TempDID    │  │
+│  │  fct: AES-GCM    │  │
+│  │  att: []         │  │
+│  │  prf: ...        │  │
+│  │                  │  │
+│  └──────────────────┘  │
+│                        │
+└────────────────────────┘
+```
+
```
Comment (dholms) [inline README.md:198]: Is there a concern here that some attacker continuously responds to messages before the actual responder?

I'm specifically thinking about regenerating the temp DID here.
- Alice sends out request which is seen by Bob and Eve.
- Eve quickly responds with a bad UCAN
- Alice throws out tempDID & tries again
- repeat until Bob gets a faster internet connection 😛

File: README.md [hunk 7]
```diff
@@ -1 +1,457 @@
...
+The Responder MUST generate a fresh 256-bit AES key and 12-byte initialization vector (IV) per connection request. It is RECOMMENDED that the Responder track all DIDs requested with, and to only respond to each temporary DID exactly once.
+
+The key used in this step MUST be used as input key material (IKM) to derive keys in rest of the AWAKE bootstrap. The AES key MUST be encoded as padded base64 and included in the facts (`fct`) section of the payload UCAN. The rest of the [validation UCAN is constructed](#332-validation-ucan) and signed per normal, thus including the AES key in the signature payload. This signature serves as proof that the AES key was intended for this specific session.
+
+``` javascript
+{
+  fct: [
+    {
+      "awake/ikm": base64PaddedIkm,
```
Comment (dholms) [inline README.md:211]: Why is this necessary if you already need the key to decrypt it?

File: README.md [hunk 8]
```diff
@@ -1 +1,457 @@
...
+``` javascript
+{
+  ...,
+  "fct": [
+    {"awake/challenge": "oob-pin"}
+  ]
+}
+```
+
+To set the challenge as `ucan`, the `fct` section of the UCAN MUST include the following:
```
Comment (dholms) [inline README.md:240]: Oh this is interesting. I like that

File: README.md (hunk 7 above)
Comment (expede) [reply README.md]: There's a reason, but this step is changing substantially due to the double ratchet so 🤷‍♀️

File: README.md (hunk 6 above)
Comment (expede) [reply README.md]: Oh yeah good point, I should add text that this is nonblocking 👍 (i.e. multiple connections allowed)

File: README.md (hunk 5 above)
Comment (expede) [reply README.md:83]: Good questions!

> Is the channel assumed to be long standing?

Yes, it's assumed that it will last at least as long as the handshake

> If not, how often does the requestor broadcast intent?

Whenever they want to start a session (during a session they may exchange a secret to speed up future connections, but not in the protocol)

> If the provider joins after the requestor broadcasts intent, when does it get the update that it needs to re-broadcast?

Yup!

**Comment by expede** [comment]
@dholms & @bgins: okay this is RFR now 🙌 

Buuut also it's Friday. Please ignore this message until at least Monday 😉

**Comment by expede** [comment]
@QuinnWilton if you could give this a look over with your CTF hat on, I'd love to know if anything jumps out at you as being horribly broken 🙏

File: README.md [hunk 9]
```diff
@@ -1 +1,649 @@
...
+
+Capability-based systems have a helpful philosophy towards a third path. By emphasizing authorization over authentication, they provide a way to know something provable about what the other party "can do", even if they have no sure way of knowing "who they are". One way of phrasing this is that such an agent is "functionally equivalent to the principal in this context". AWAKE makes use of authorization to bootstrap point-to-point sessions that are both secure and mutually trusted.
+
+## 1.1 Terminology
+
+This document contains shorthand (especially in diagrams) and nuanced senses of some terms. Below is a dictionary of AWAKE-specific terms:
+
+| Term      | Meaning                                              |
+| --------- | ---------------------------------------------------- |
+| Attacker  | An attacker attempting to gain access to the channel |
```
Comment (QuinnWilton) [inline README.md:32]: Really nitpicky comment, but is it worth avoiding `attacker` in the definition? Just something like:
> A malicious third party attempting to gain access to the channel

File: README.md [hunk 10]
```diff
@@ -1 +1,649 @@
...
+| PK        | Public key                                           |
+| Requestor | The agent opening the session                        |
+| Responder | The agent being contacted by the Requestor           |
+| SK        | Secret (private) key                                 |
+
+## 1.2 Payload Fields
+
+Payloads are encoding agnostic, but JSON is RECOMMENDED. For JSON, any fields that contain non-JSON values (such as ECDH public keys and encryption payloads) MUST be serialized as unpadded [Base64](https://datatracker.ietf.org/doc/html/rfc4648).
+
+Messages that a peer cannot parse SHOULD be ignored.
```
[the remark this answers is not in the document]
Comment (expede) [reply README.md:43]: Yeah, I can remove that line. The `RECOMMENDED` is there because it's not a _hard_ requirement (e.g. maybe you store-and-forward unreadable messages). Mainly I wanted to convey that it's fine to ignore messages that you can't parse because they're either not meant for you or malicious.

Would you just remove this line?

[the remark this answers is not in the document]
Comment (QuinnWilton) [reply README.md:43]: Makes sense! I'd probably either remove it, or change it to `MUST`. Someone can still do whatever they want with the messages, outside of the flow of the protocol, but they shouldn't be processed as part of AWAKE.

File: README.md [hunk 11]
```diff
@@ -1 +1,649 @@
...
+```javascript
+{
+  "did": requestorActualDid,
+  "sig": signedHash
+}
+```
+
+#### 3.4.2.3 Direct UCAN Challenge
+
+If UCAN auth is required by the Responder, the Requestor MUST provide a UCAN. This is the same strategy as the one used by the Responder in [§3.3](#33-responder-es tablishes-point-to-point-session): the UCAN MUST be encrypted with the session key and the IV from the enclosing payload, MUST be given in a raw format, and MUST be inline (without a JSON object wrapper or similar).
```
Comment (QuinnWilton) [inline README.md:454]: The link to `§3.3` is broken here

File: README.md [hunk 12]
```diff
@@ -1 +1,649 @@
...
+| ------ | --------------------------------- | -------------------------------------------------------------- | -------- |
+| `awv`  | `"0.1.0"`                         | AWAKE message version                                          | Yes      |
+| `type` | `"awake/msg"`                     | Generic AWAKE message type                                     | Yes      |
+| `id`   | `sha3_256(resEcdhPk + reqEcdhPk)` | Message ID                                                     | Yes      |
+| `iv`   |                                   | Initialization vector for the encrypted payload                | Yes      |
+| `msg`  |                                   | Fulfilled challenge payload encrypted with Step 4 ECDH AES-key | Yes      |
+
+#### 3.5.1.1 Encrypted Message
+
+The encrypted message payload MUST include an `awake/ack` field, with a value of the requestor's long-term DID. This payload MAY contain additional fields. This is often useful if dovetailing the ACK with the first message of a session using the
```
Comment (QuinnWilton) [inline README.md:500]: Did this last sentence get cut off?

File: README.md [hunk 13]
```diff
@@ -1 +1,649 @@
...
+| `mid`  | `sha3_256(latestSenderEcdhPk + latestReceiverEcdhPk)` | Message ID                                                              | Yes      |
+| `sid`  | `sha3_256(firstAesKey)`                               | OPTIONAL session ID                                                     | No       |
+| `iv`   |                                                       | Initialization vector for the encrypted payload                         | Yes      |
+| `msg`  |                                                       | Fulfilled challenge payload encrypted with latest KDF AES-key           | Yes      |
+| `esig` | `encrypt(messageKey, sig(senderUcanSk, msg + mid))`   | OPTIONAL message signature using the sender's UCAN-validated public key | No       |
+
+The OPTIONAL encrypted signature field (`esig`) MAY be included to prove authenticity of the payload. Without this field, an eavesdropper with access to the input secret MAY be able to spoof a message that updates the next ECDH key to one it controls. The signature MUST be encrypted, since some commonly used key types such as RSA allow for the derivation of the signer's public key.
```
Comment (QuinnWilton) [inline README.md:533]: I'm having a little bit of trouble visualizing this attack. Do you have a situation in mind where an eavesdropper will have access to the input secret, but without having compromised the device in a way that allows them to produce this signature themselves?

I ask because the `esig` field being optional raises some alarms for me: this is the sort of thing implementers would skip in a first version, so if it's an important mitigation, maybe it shouldn't be optional. And if it actually isn't required, it might be helpful to include a little bit more guidance on making that decision.

File: README.md [hunk 14]
```diff
@@ -1 +1,649 @@
...
+Requestor                  Responder
+    ⋮                          ⋮
+    │                          │
+    │           FIN            │ (7)
+    │◄────────────────────────►│
+    │                          │
+    ▀                          ▀ 
+```
+
+Graceful disconnection from an AWAKE attempt MAY be broadcast at any time during an AWAKE session, including to cancel the AWAKE handshake attempt. This payload MUST be the standard [AWAKE message type](#4-secure-session), SHOULD NOT contain any other keys.
```
Comment (QuinnWilton) [inline README.md:583]: nitpick: add "and" before "SHOULD NOT"

> This payload MUST be the standard [AWAKE message type](#4-secure-session), and SHOULD NOT contain any other keys.

File: README.md [hunk 15]
```diff
@@ -1 +1,649 @@
...
+
+All symmetric encryption in AWAKE MUST use [256-bit AES-GCM](https://csrc.nist.gov/publications/detail/sp/800-38d/final). These keys MUST be generated by the [KDF](#143-key-derivation), and SHOULD be non-extractable where possible.
+
+Each encrypted payload MUST include a unique (freshly generated) 12-byte [initialization vector](https://en.wikipedia.org/wiki/Initialization_vector).
+
+### 1.4.3 Key Derivation
+
+AWAKE uses a modified [Extended Triple Diffie-Hellman (X3DH)](https://www.signal.org/docs/specifications/x3dh/) protocol in order to support the WebCrypto API's non-extractable keys and make use of existing UCAN tokens without the need for pre-sharing keys.
+
+Key derivation in AWAKE's double ratchet MUST use the following algorithm:
```
Comment (QuinnWilton) [inline README.md:74]: Is there a reason you're using SHA3 to derive the OKM directly, rather than going through something like [HKDF](https://developer.mozilla.org/en-US/docs/Web/API/HkdfParams)?

I notice that the spec doesn't give any guidance on generating the IV, but with HKDF you'd be able to generate both the OKM and the IV from the output of `ECDH`, and avoid needing to send it along through the protocol. You'd just need to make sure to use the `info` parameter of HKDF to guarantee domain separation for both derivations.

File: README.md [hunk 16]
```diff
@@ -1 +1,649 @@
...
+    │                          │
+    │         Messages         │ (6)
+    │◄────────────────────────►│
+    │                          │
+    ⋮                          ⋮
+```
+
+Messages sent over an established AWAKE session MUST contain the following keys:
+ 
+| Field  | Value                                                 | Description                                                             | Required |
```
Comment (QuinnWilton) [inline README.md:523]: Nitpick: this table has both a `Required` column, and you say `OPTIONAL` in the descriptions of any non-required fields.

File: README.md [hunk 17]
```diff
@@ -1 +1,649 @@
...
+| ------ | ----------------------------------------------------- | ----------------------------------------------------------------------- | -------- |
+| `awv`  | `"0.1.0"`                                             | AWAKE message version                                                   | Yes      |
+| `type` | `"awake/msg"`                                         | Generic AWAKE message type                                              | Yes      |
+| `mid`  | `sha3_256(latestSenderEcdhPk + latestReceiverEcdhPk)` | Message ID                                                              | Yes      |
```
Comment (QuinnWilton) [inline README.md:527]: Is it worth linking to the relevant sections for "Message ID" and "Session ID" in this and the other tables?

File: README.md [hunk 18]
```diff
@@ -1 +1,629 @@
...
+    {"awake/nextdid": step4EcdhDid}
+  ]
+}
+```
+
+If more than one `awake/nextdid` field is set, the lowest-indexed one MUST be used.
+
+### 3.3.2 Payload
+
+To start the Double Ratchet, the payload in this stage has the highest number of cleartext fields. Note that the value in the `res` field contain the temporary ECDH DIDs, and MUST NOT use the Responder's actual long-term DID. Conversely, the UCAN inside the encrypted payload MUST use the Responder's long-term DID.
```
Comment (bgins) [inline README.md:371]: > Note that the value in the `res` field contain the temporary ECDH DIDs

Should `res` be the `iss`? I see `res` in the message type, but not as a field.

File: README.md (hunk 11 above)
Comment (expede) [reply README.md:454]: Oh I see. The link checker probably returns a `200`, even if that specific anchor is broken. Welp, I guess back to manually checking. Thanks!

File: README.md [hunk 19]
```diff
@@ -1 +1,649 @@
...
+
+## 7.2 IKEv2
+
+The [Internet Key Exchange (IKE) Protocol](https://datatracker.ietf.org/doc/html/rfc7296) is typically (but not exclusively) used as part of [IPsec](https://en.wikipedia.org/wiki/IPsec). IKE generally uses certificate authorities (CAs). IKE requires that X.509 be supported.
+
+IKE shares many commonalities with AWAKE, including making available of the same cryptographic algorithms (e.g. P-256).
+
+## 7.3 WireGuard
+
+[WireGuard](https://www.wireguard.com/) is a VPN protocol that is widely deployed via the Linux kernel, and has since been ported to many other systems. It is UDP-based and aimed at raw performance and security. Being so level, is unconstrained in which cryptographic primitives is uses (i.e. Curve25519).
```
Comment (bgins) [inline README.md:636]: > Being so level, is unconstrained in which cryptographic primitives is uses (i.e. Curve25519).

Should this sentence open with "Being low-level, it is..."? May be something else, but I'm guessing that might be it because it's in the kernel.

File: README.md (hunk 15 above)
Comment (expede) [reply README.md:74]: HKDF was used in an earlier draft, because it's available in the WebCrypto API. The short answer is that we can totally use it if it's better. You've gone deeper on this stuff than me, so I'd love to explain the reasoning and get your thoughts!

# What Does Signal Do?

Signal's Double Ratchet paper suggests two systems for the KDF: HMAC (SHA) and HKDF.

# In Favour of HKDF

HKDF intends to be an easy choice to "just use" as a generic KDF. The people that designed it are -- frankly -- much better at cryptography than me!

# Against HKDF

On the flip side, hash chaining HMACs to derive keys is pretty widely used.

HKDF's security comes from SHA2 which is significantly slower than SHA3. HKDF is much more general than a hash function, and does things like expansion:

> 2. To "expand" the generated output of an already reasonably random input such as an existing shared key into a larger cryptographically independent output
> — Wikipedia

In our use case, we need exactly 256-bit outputs, so features like expansion (and the additional computation therein) isn't required. Breaking a SHA3-derived key is the same as reversing the hash, which "should" be intractable (under hash assumptions).

File: README.md (hunk 18 above)
Comment (expede) [reply README.md:371]: It should! Thanks :)

File: README.md (hunk 19 above)
Comment (expede) [reply README.md]: Indeed! Fixed 🙏

File: README.md (hunk 16 above)
Comment (expede) [reply README.md:523]: Fair! ✂️

File: README.md (hunk 13 above)
Comment (expede) [reply README.md:533]: haha I had a feel that this was going to come up 😅 

The scenario is that if an attacker manages to spoof a single message, they can _change_ the next ECDH key. Signing with the UCAN DID key avoids this because it authenticates the message without depending on the AES envelope. Signing can't hurt, but also break-in security is the point of the double ratchet.

That said, when writing this section, it did feel like a bridge too far / like I was wearing a stack of tinfoil hats.

You're 100% right that hedging with a SHOULD is the wrong strategy here.

I'll go ahead and drop this field. Thanks for the rubber ducking 🦆

File: README.md (hunk 11 above)
Comment (expede) [reply README.md:454]: We wouldn't have this problem with LaTeX 🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔

File: README.md (hunk 15 above)
Comment (QuinnWilton) [reply README.md:74]: > You've gone deeper on this stuff than me, so I'd love to explain the reasoning and get your thoughts!

This also definitely isn't my expertise, and most of what I know is just sourced from cryptographers with far more experience than me!

I don't believe there's anything insecure about using SHA3 here, like you are, but I've got two comments.

> HKDF's security comes from SHA-2 which is significantly slower than SHA-3. 

This sounds backwards to me? I thought SHA-3 was widely considered to be slower, except on specialized hardware. This is a point that's brought up by [Adam Langley](https://www.imperialviolet.org/2017/05/31/skipsha3.html):

> SHA-3 is also [slow](http://bench.cr.yp.to/results-sha3.html), and is even slower than SHA-2 which is already a comparative laggard amongst crypto primitives.

And is also something [David Wong acknowledges](https://cryptologie.net/article/400/maybe-you-shouldnt-skip-sha-3/). He concludes that you should just choose SHA-3 as a general-purpose hashing algorithm anyway, because avoiding the risk of length extension attacks are worth the slowdown, but that's not something we need to worry about with HKDF.

I also [found some hashcat benchmarks](https://gist.github.com/Chick3nman/ccfb883d2d267d94770869b09f5b96ed) someone ran on an M1 earlier this year, and it looks like SHA-2 outperforms SHA-3 there:

```
---------------------------
* Hash-Mode 1400 (SHA2-256)
---------------------------

Speed.#1.........:  1818.9 MH/s (54.07ms) @ Accel:128 Loops:128 Thr:128 Vec:1

----------------------------
* Hash-Mode 17400 (SHA3-256)
----------------------------

Speed.#1.........:   581.7 MH/s (85.14ms) @ Accel:128 Loops:64 Thr:128 Vec:1
```

> In our use case, we need exactly 256-bit outputs, so features like expansion (and the additional computation therein) isn't required.

I'd actually argue that we need 352-bit outputs, because we have a 256-bit key, and a 96-bit IV. Whether or not it's generated as part of the KDF, that IV is needed at some point, so implementers will need to find another 12-bytes each round from somewhere, and I'd argue that baking the generation of the IV into the KDF would both limit possible mistakes elsewhere (using the same IV in both directions, for example).

> I'm weary of pseudorandom IVs, but maybe that's just me?

My understanding of the IV in AES-GCM is that it just needs to be unique, so one of my concerns is that someone will attempt to do so by either using a hardcoded IV (because each party only sends a single message per key), or by using a counter. In both cases they'll be forgetting to take into account that the other party also sends a message, and so would be making use of the same IV.

My thought with using HKDF to derive the IV was to use the sender's DID as the salt to HKDF so that both parties are able to use the same key material to generate a unique IV per round.

> There's a few places in AWAKE that we could be more economical with space, but ease of implementation is the bigger concern for this version.

That makes sense! I agree that saving space isn't the highest priority here, and the larger point behind the suggestion was about removing a degree of freedom from implementers that may lead to IV misuse, with removing that field just being a bonus.

Either way though, this comment definitely isn't me saying that you're doing anything wrong and that it needs to change or anything like that! I was mostly curious about what had led you toward SHA-3 in the first place.

Comment (expede) [reply README.md:74]: > This sounds backwards to me? I thought SHA-3 was widely considered to be slower, except on specialized hardware

We ran some benchmarks roughly last autumn to test out Wasm, and IIRC the Wasm SHA3-256 outperformed the WebCrypto API's SHA2-256 (not an apples-to-apples comparison, but you'd expect the native function to outperform Wasm). IIRC the SHA3 team also believes this to be the case for pure software implementations.

But that said, we control the stack here, so SHA2 or BLAKE3 or similar are all on the table!

Lemme dig up that post about SHA2 vs SHA3, though.

Comment (expede) [reply README.md:74]: > My understanding of the IV in AES-GCM is that it just needs to be unique

Okay, yeah, that's true!

> In both cases they'll be forgetting to take into account that the other party also sends a message, and so would be making use of the same IV.

That is a _really good point_ 💯

Comment (QuinnWilton) [reply README.md:74]: > We ran some benchmarks roughly last autumn to test out Wasm, and IIRC the Wasm SHA3-256 outperformed the WebCrypto API's SHA2-256

Fascinating! I'm seeing the same thing echoed in other benchmarks, like [this](https://github.com/bitauth/libauth/wiki/Benchmarks#browser-sha256-hash-a-32-byte-input) `libauth` benchmark that puts their Wasm sha256 implementation at almost 20x the throughput of `crypto.subtle`, for 32-byte inputs:

```
libauth x 571,012 ops/sec ±2.38% (62 runs sampled)
hash.js x 346,025 ops/sec ±1.31% (63 runs sampled)
crypto.subtle x 37,260 ops/sec ±2.15% (58 runs sampled)
asmcrypto.js x 3,138 ops/sec ±2.38% (20 runs sampled)
```

File: README.md [hunk 20]
```diff
@@ -1 +1,634 @@
...
+# Authorized Wire Authenticated Key Exchange (AWAKE) Specification v0.1.0
+
+## Editors
+
+* [Brooklyn Zelenka](https://github.com/expede), [Fission](https://fission.codes)
+
+## Authors
+
+* [Daniel Holmgren](https://github.com/dholms), [Bluesky](https://blueskyweb.xyz/)
+* [Quinn Wilton](https://github.com/QuinnWilton), [Fission](https://fission.codes)
```
Comment (expede) [inline README.md:10]: cc @QuinnWilton

File: README.md [hunk 21]
```diff
@@ -1 +1,638 @@
...
+                                  Next
+                                 Secret
+```
+
+``` javascript
+// JS-flavored Pseudocode
+
+const ecdhSecret = ecdh(aliceSk, bobPk)
+const awakeTag = 0x4157414B452D5543414E // "AWAKE-UCAN" as hex
+const pseudorandomBits = hkdf.generateBits({ecdh, salt: currentSecret, info: awakeTag, bitLength: 256 + 256 + 12})
```
Comment (QuinnWilton) [inline README.md:116]: Should this be `256 + 256 + 96`?

Comment (expede) [reply README.md:116]: How different are bits and bytes _really?_

Comment (QuinnWilton) [reply README.md:116]: 7 bits

Comment (expede) [reply README.md:116]: You're not wrong

File: README.md (hunk 11 above)
Comment (QuinnWilton) [reply README.md:437]: But only because we'd be too busy fumbling around with spacing.

image

**Review by bgins — APPROVED** [review]
editor

Looks great! The new double ratchet is fantastic and a nice expansion of the protocol. I left a few copy edits throughout.

One question about the double ratchet. My understanding is that at each step a new set of key pair components are used to create a new secret. With each step, a single message can be sent. It seems like each step blocks the next because a new key pair component is needed to progress. How can there be out of order messages? Wouldn't the algorithm block waiting to produce the next secret?

**Comment by expede** [comment]
> One question about the double ratchet. My understanding is that at each step a new set of key pair components are used to create a new secret. With each step, a single message can be sent. It seems like each step blocks the next because a new key pair component is needed to progress. How can there be out of order messages? Wouldn't the algorithm block waiting to produce the next secret?

@bgins Yes this is the case during the handshake, because that must proceed in lock-step. After The handshake (during the main message session), this requirement is relaxed so that a sender can produce a bunch of messages in one go without having to ratchet their ECDH key. The keys are then run through a SHA512 to get the next AES key and IV. These parts give up future secrecy until the ECDH chain is ratcheted.

Basically, every time you send a `awake/nextdid`, new sending/receiving chains get initialized from scratch (for future secrecy), but if an attacker breaks into one of the messages between ECDH steps, they'll be able to read from that point until before the next ECDH step.

Do you have any thoughts on which parts could be made clearer and/or more salient to the next reader?

File: README.md [hunk 22]
```diff
@@ -1 +1,726 @@
...
+│  │  att: []          │  │
+│  │  fct: nextResECDH │  │
+│  │  prf: ...         │  │
+│  │                   │  │
+│  └───────────────────┘  │
+│                         │
+└─────────────────────────┘
+```
+
+Upon receipt, the Requestor MUST validate that the UCAN capabilities fulfill their `caps` criteria. The UCAN itself MUST be valid, unrevoked, unexpired, and intended for the temporary DID (the `aud` field). If any of these checks fail, the session MUST be abandoned, the temporary DID regenerated, and the protocol restarted from [intention broadcast](#32-requestor-broadcasts-intent).
```
[the remark this answers is not in the document]
Comment (expede) [reply README.md:402]: Good suggestion, thanks!

@bgins I recently learned about the new GitHub `suggestion` feature, which may be a handy tool for you too:

```suggestion
Upon receipt, the Requestor MUST validate that the UCAN capabilities in the proof fulfill their caps criteria.
```

Just use `suggestion` as the tag for a multi-line code snippet, just like how you would tripple-backtick and put `js`

```js
const foo = bar
```

File: README.md [hunk 23]
```diff
@@ -1 +1,726 @@
...
+
+### 3.3.1 Validation UCAN
+
+The validation UCAN MUST NOT be used to delegate any capabilities. This UCAN MUST only be used to prove access to capabilities and sign the AES key. The `att` and `my` fields MUST be empty arrays. The issuer (`iss`) field MUST contain the Responder's long-term DID (rather than the temporary ECDH DID). The audience (`aud`) field MUST contain the Requestor's temporary ECDH DID from [§3.2](#32-requestor-broadcasts-intent).
```
[the remark this answers is not in the document]
Comment (expede) [reply README.md:406]: Good catch! Indeed, this is from the previous KEK design 👍

File: README.md [hunk 24]
```diff
@@ -1 +1,726 @@
...
+  "did": requestorActualDid,
+  "sig": signedHash
+}
+```
+
+#### 3.4.2.3 Direct UCAN Challenge
+
+If UCAN auth is required by the Responder, the Requestor MUST provide a UCAN. This is the same strategy as the one used by the Responder in [§3.3](#33-responder-es tablishes-point-to-point-session): the UCAN MUST be encrypted with the session key and the IV from the enclosing payload, MUST be given in a raw format, and MUST be inline (without a JSON object wrapper or similar).
+
+The UCAN MUST be issued (`iss`) by the Requestor's DID (not the temporary DID), and its audience (`aud`) MUST be the Responder's DID. The `att` field MUST be set to an empty array (i.e. it MUST NOT delegate any capabilities). The `prf` array MUST fulfill the capabilities set by the Responder.
```
[the remark this answers is not in the document]
Comment (expede) [reply README.md:543]: @bgins How about the newly added text just below on line 545?

File: README.md (hunk 22 above)
[the remark this answers is not in the document]
Comment (bgins) [reply README.md]: Hey, that's great! I'll give that a try on my next suggestion. 😄

**Comment by bgins** [comment]
> Do you have any thoughts on which parts could be made clearer and/or more salient to the next reader?

This part of section 4.1 could be clearer:

https://github.com/ucan-wg/awake/blob/320235bc65857d84aeff978d2f693ea1fab2f334/README.md?plain=1#L622

The first sentence makes it seem like each message needs some unique key pair component, and they need to ratchet forward so that a secret is only used in two messages. 

Maybe this section could be updated to account for message sequences that use the same secret, but increment the message count? I think the message count is what would make the messages in a sequence unique?
